clang/include/clang/ExtractAPI/Serialization/SerializerBase.h

```diff
@@ -28,11 +28,78 @@ struct APISerializerOption {
 };
 
 /// The base interface of serializers for API information.
-class APISerializer {
+template <typename Derived> class APISerializer {
 public:
   /// Serialize the API information to \p os.
   virtual void serialize(raw_ostream &os) = 0;
 
+  virtual void traverseGlobalFunctionRecords() {
+    for (const auto &globalFunction : API.getGlobalFunctions())
+      static_cast<Derived *>(this)->visitGlobalFunctionRecord(
+          std::move(*globalFunction.second));
+  }
+
+  void traverseGlobalVariableRecords() {
+    for (const auto &global : API.getGlobalVariables())
+      static_cast<Derived *>(this)->visitGlobalVariableRecord(*global.second);
+  }
+
+  void traverseEnumRecords() {
+    for (const auto &enum_nonkeyword : API.getEnums())
+      static_cast<Derived *>(this)->visitEnumRecord(*enum_nonkeyword.second);
+  }
+
+  void traverseStructRecords() {
+    for (const auto &Struct : API.getStructs())
+      static_cast<Derived *>(this)->visitStructRecord(*Struct.second);
+  }
+
+  void traverseObjCInterfaces() {
+    for (const auto &interface : API.getObjCInterfaces())
+      static_cast<Derived *>(this)->visitObjCContainerRecord(*interface.second);
+  }
+
+  void traverseObjCProtocols() {
+    for (const auto &protocol : API.getObjCProtocols())
+      static_cast<Derived *>(this)->visitObjCContainerRecord(*protocol.second);
+  }
+
+  void traverseMacroDefinitionRecords() {
+    for (const auto &definition : API.getMacros())
+      static_cast<Derived *>(this)->visitMacroDefinitionRecord(
+          *definition.second);
+  }
+
+  void traverseTypedefRecords() {
+    for (const auto &Typedef : API.getTypedefs())
+      static_cast<Derived *>(this)->visitTypedefRecord(*Typedef.second);
+  }
+  /// Visit a global function record.
+  virtual void
+  visitGlobalFunctionRecord(const GlobalFunctionRecord &Record) = 0;
+
+  /// Visit a global variable record.
+  virtual void
+  visitGlobalVariableRecord(const GlobalVariableRecord &Record) = 0;
```

Comment:
Again using the CRTP means that you don't need to mark these as virtual. The accepted way of doing this would be to provide an empty implementation as follows:
```c++
void visitGlobalVariableRecord(const GlobalVariableRecord &Record) { }
```